p2p/src/peer_manager/peerdb/mod.rs

```diff
@@ -166,7 +171,15 @@ where
                     None
                 }
             })
-            .choose_multiple(&mut make_pseudo_rng(), count)
+            .choose_multiple(&mut make_pseudo_rng(), count);
+
+        // Drop duplicate address groups as needed (shuffle selected addresses first to make the selection fair)
+        selected.shuffle(&mut make_pseudo_rng());
+        let selected_grouped = selected
+            .into_iter()
+            .map(|a| (get_address_group(&a.as_peer_address()), a))
+            .collect::<HashMap<_, _>>();
+        selected_grouped.values().cloned().collect()
```

Comment:
You can use `into_values()` instead of `values().cloned()` to avoid the extra cloning.

But a better solution is to just use `unique_by(|a| get_address_group(&a.as_peer_address()))`, and avoid the whole collecting into a HashMap.

Reply:
Nice, thanks!